core/types/calls.cc

```diff
@@ -137,6 +137,53 @@ DispatchResult TupleType::dispatchCall(const GlobalState &gs, const DispatchArgs
     return dispatchCallProxyType(gs, underlying(gs), args);
 }
 
+DispatchResult SelfTypeParam::dispatchCall(const GlobalState &gs, const DispatchArgs &args) const {
+    if (this->definition.isTypeArgument()) {
+        auto result = DispatchResult(Types::untypedUntracked(), std::move(args.selfType), Symbols::noMethod());
+        if (args.suppressErrors) {
+            // Short circuit here to avoid constructing an expensive error message.
+            return result;
+        }
+        auto e = gs.beginError(args.callLoc(), errors::Infer::UnknownMethod);
+        if (e) {
+            auto thisStr = args.thisType.show(gs);
+            if (args.fullType.type != args.thisType) {
+                e.setHeader("Method `{}` does not exist on `{}` component of `{}`", args.name.show(gs), thisStr,
+                            args.fullType.type.show(gs));
+            } else {
+                e.setHeader("Call to method `{}` on unconstrained generic type `{}`", args.name.show(gs), thisStr);
+            }
+
+            auto receiverLoc = core::Loc{args.locs.file, args.locs.receiver};
+            if (receiverLoc.exists() && (gs.suggestUnsafe.has_value())) {
+                auto wrapInFn = gs.suggestUnsafe.value();
+                if (receiverLoc.empty()) {
+                    auto shortName = args.name.shortName(gs);
+                    auto beginAdjust = -2;                     // (&
+                    auto endAdjust = 1 + shortName.size() + 1; // :foo)
+                    auto blockPassLoc = receiverLoc.adjust(gs, beginAdjust, endAdjust);
+                    if (blockPassLoc.exists()) {
+                        auto blockPassSource = blockPassLoc.source(gs).value();
+                        if (blockPassSource == fmt::format("(&:{})", shortName)) {
+                            e.replaceWith(fmt::format("Expand to block with `{}`", wrapInFn), blockPassLoc,
+                                          " {{|x| {}(x).{}}}", wrapInFn, shortName);
+                        }
+                    }
+                } else {
+                    e.replaceWith(fmt::format("Wrap in `{}`", wrapInFn), receiverLoc, "{}({})", wrapInFn,
+                                  receiverLoc.source(gs).value());
+                }
+            }
+        }
+        result.main.errors.emplace_back(e.build());
+        return result;
+    } else {
+        // TODO: https://github.com/sorbet/sorbet/issues/1731
+        auto untypedUntracked = Types::untypedUntracked();
+        return untypedUntracked.dispatchCall(gs, args.withThisRef(untypedUntracked));
+    }
```

Comment:
@elliottt question for you: Do you have an intuition for why `SelfTypeParam` models both `type_member` symbols and `type_parameter` symbols?

There's some code over here in `isSubTypeUnderConstraintSingle` that seems to do something similar (different logic depending on type argument vs type member):

https://github.com/sorbet/sorbet/blob/96c4d8e18db19a677b96e3edc41bcd9fa1365781/core/types/subtyping.cc#L1096-L1129

Reply:
I've always thought of `SelfTypeParam` as a skolem variable, and using it to represent both a `type_member` and `type_parameter` is probably just for convenience.

In this case in subtyping.cc, we're using the fact that `LambdaParam` has bounds information present to discover cases where we know they must be compatible. In all other cases they're considered to be incompatible. [Here's an example](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Aclass%20Parent%3B%20end%0Aclass%20Child%20%3C%20Parent%3B%20end%0A%0Aclass%20WithBounds%0A%20%20extend%20T%3A%3AGeneric%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20A%20%3D%20type_member%28upper%3A%20Child%29%0A%20%20B%20%3D%20type_member%28lower%3A%20Parent%29%0A%0A%20%20sig%20%7Breturns%28A%29%7D%0A%20%20def%20make_A%0A%20%20%20%20T.unsafe%28nil%29%0A%20%20end%0A%0A%20%20sig%20%7Bparams%28b%3A%20B%29.void%7D%0A%20%20def%20uses_B%28b%29%0A%20%20end%0A%0A%20%20def%20combine%0A%20%20%20%20a%20%3D%20self.make_A%0A%20%20%20%20self.uses_B%28a%29%0A%20%20end%0A%0Aend%0A%0Aclass%20WithoutBounds%0A%20%20extend%20T%3A%3AGeneric%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20A%20%3D%20type_member%28%29%0A%20%20B%20%3D%20type_member%28%29%0A%0A%20%20sig%20%7Breturns%28A%29%7D%0A%20%20def%20make_A%0A%20%20%20%20T.unsafe%28nil%29%0A%20%20end%0A%0A%20%20sig%20%7Bparams%28b%3A%20B%29.void%7D%0A%20%20def%20uses_B%28b%29%0A%20%20end%0A%0A%20%20def%20combine%0A%20%20%20%20a%20%3D%20self.make_A%0A%20%20%20%20self.uses_B%28a%29%0A%20%20end%0A%0Aend) where the bounds information allows interactions of `type_member` values in a context where they're skolemized.